android/greengrass-android-component-base/component-library/src/main/java/com/aws/greengrass/android/component/core/GreengrassComponentActivity.java

```diff
@@ -0,0 +1,162 @@
+public abstract class GreengrassComponentActivity extends AppCompatActivity {
+
+    // will true since START cmd is received and till end of stop(), used to prevent double start
+    private AtomicBoolean isStarted = new AtomicBoolean(false);
+
+    /* will false since end of start() and till just before any stop() attempt,
+         used to prevent double stop */
+    private AtomicBoolean isStopped = new AtomicBoolean(true);
+
+    // Name of master package to send responses to
+    private String masterPackage = null;
+
+    // Component's arguments
+    private String[] arguments = null;
+
+    // environment
+    private AtomicReference<HashMap<String, String>> environment
+            = new AtomicReference<HashMap<String, String>>(new HashMap<String, String>());
+
+    private final BroadcastReceiver receiver = new BroadcastReceiver() {
+        @Override
+        public void onReceive(Context context, Intent intent) {
+            try {
+                String action = intent.getAction();
+                if (!isStopped.getAndSet(true)) {
+                    stop(EXIT_CODE_TERMINATED);
+                }
+            } catch (Throwable e) {
+                // TODO: log message
+                e.printStackTrace();
+            }
+        }
+    };
+
+    /**
+     * Gets the value of a single environmental variable.
+     *
+     * @param name The name of the variable to retrieve
+     * @return A string representing the value of the specified environmental variable,
+     *      null if the specified variable does not exist.
+     */
+    @Nullable
+    protected String getEnvironmentalVariable(final String name) {
+        String value = null;
+        HashMap<String, String> env = environment.get();
+        if (env != null) {
+            value = env.get(name);
+        }
+        return value;
+    }
+
+    /**
+     * Gets component's arguments.
+     *
+     * @return array of component's arguments.
+     */
+    protected String[] getArguments() {
+        return arguments;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    protected void onCreate(Bundle savedInstanceState) {
+        try {
+            super.onCreate(savedInstanceState);
+            registerBroadcastReceiver();
+            if (ACTION_START_COMPONENT.equals(getIntent().getAction())) {
+                if (!isStarted.getAndSet(true)) {
+                    Intent intent = getIntent();
+                    masterPackage = intent.getStringExtra(EXTRA_MASTER_PACKAGE);
+                    arguments = intent.getStringArrayExtra(EXTRA_ARGUMENTS);
+
+                    // Try to extract Environment supplied by Nucleus
+                    if (intent.hasExtra(EXTRA_COMPONENT_ENVIRONMENT)) {
+                        HashMap<String, String> env =
+                                (HashMap<String, String>) intent.getSerializableExtra(EXTRA_COMPONENT_ENVIRONMENT);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Use `Parcelable`, not `Serializable`, to pass data between different components in Android. `Parcelable` is more performant. See the following diff on GitHub that makes this code change: [diff](https://github.com/k9mail/k-9/commit/685df0ff6772df853457e782809ab3d064f533d5#diff-3e4082b13ee86b8238eff49b864ff82db653e5af02ce4ee9c7e21b6c9bfca3a2L330)

Learn more: [Parcelable in the Android API reference](https://developer.android.com/reference/android/os/Parcelable) and [Serializable and Parcelable in Android](https://skoric.svbtle.com/serializable-vs-parcelable)